Bio/AlignIO/MafIO.py

```diff
@@ -0,0 +1,581 @@
+# Revisions Copyright 2014, 2015 by Adam Novak.  All rights reserved.
+#
+# This code is part of the Biopython distribution and governed by its
+# license.  Please see the LICENSE file that should have been included
+# as part of this package.
+"""Bio.AlignIO support for the "maf" multiple alignment format.
+
+You are expected to use this module via the Bio.AlignIO functions(or the
+Bio.SeqIO functions if you want to work directly with the gapped sequences).
+
+"""
+from Bio.Alphabet import single_letter_alphabet
+from Bio.Seq import Seq
+from Bio.SeqRecord import SeqRecord
+from Bio.Align import MultipleSeqAlignment
+from .Interfaces import SequentialAlignmentWriter
+import shlex
+
+MAFINDEX_VERSION = 1
+
+
+class MafWriter(SequentialAlignmentWriter):
+    """Accepts a MultipleSeqAlignment object, writes a MAF file"""
+
+    def write_header(self):
+        """Writes the MAF header"""
+
+        self.handle.write("##maf version=1 scoring=none\n")
+        self.handle.write("# generated by Biopython\n\n")
+
+    def _write_record(self, record):
+        """Writes a single SeqRecord object to an 's' line in a MAF block"""
+
+        # convert biopython-style 1/-1 strand to MAF-style +/- strand
+        if record.annotations.get("strand") == 1:
+            strand = "+"
+        elif record.annotations.get("strand") == -1:
+            strand = "-"
+        else:
+            strand = "+"
+
+        fields = ["s",
+                  # In the MAF file format, spaces are not allowed in the id
+                  "%-40s" % record.id.replace(" ", "_"),
+                  "%15s" % record.annotations.get("start", 0),
+                  "%5s" % record.annotations.get("size", len(str(record.seq).replace("-", ""))),
+                  strand,
+                  "%15s" % record.annotations.get("srcSize", 0),
+                  str(record.seq)]
+        self.handle.write(" ".join(fields) + "\n")
+
+    def write_alignment(self, alignment):
+        """
+        Writes every SeqRecord in a MultipleSeqAlignment object to its own
+        MAF block (beginning with an 'a' line, containing 's' lines)
+        """
+        if not isinstance(alignment, MultipleSeqAlignment):
+            raise TypeError("Expected an alignment object")
+
+        if len(set([len(x) for x in alignment])) > 1:
+            raise ValueError("Sequences must all be the same length")
+
+        # We allow multiple sequences with the same IDs; for example, there may
+        # be a MAF aligning the + and - strands of the same sequence together.
+
+        # for now, use ._annotations private property, but restrict keys to those
+        # specifically supported by the MAF format, according to spec
+        try:
+            anno = " ".join(["%s=%s" % (x, y) for x, y in alignment._annotations.items() if x in ("score", "pass")])
+        except AttributeError:
+            anno = "score=0.00"
+
+        self.handle.write("a %s\n" % (anno,))
+
+        recs_out = 0
```

Comment:
The explicit counting is consistent with existing usage in ``Bio.SeqIO`` and ``Bio.AlignIO``, so it would be  worth updating all of them at once. But I need to be convinced this is clearer... I failed to find a Python idiom on this - so I've asked on http://stackoverflow.com/questions/42207102/python-idiom-for-counting-loop-execution